### ISSUE TITLE
Integrate WinUIEx to close platform gaps with windowing

### ISSUE THREAD

**Issue by mikebattista** [microsoft/TemplateStudio#4529]
### Is your feature request related to a problem? Please describe.
The WinUI/WinAppSDK windowing APIs aren't yet to parity with what Win32 APIs offer, however since they are HWND-based, Win32 APIs can be used together with WinUI/WinAppSDK APIs to provide a more complete set of functionality that most developers will need. We should consider ways to bridge the gap out-of-the-box with Template Studio.

### Describe the solution you'd like

https://github.com/dotMorten/WinUIEx
https://github.com/dotMorten/WinUIEx/blob/main/docs/concepts/WindowExtensions.md
https://github.com/dotMorten/WinUIEx/blob/main/docs/concepts/HwndExtensions.md
https://github.com/dotMorten/WinUIEx/blob/main/docs/concepts/WindowEx.md

Looking through the WinUIEx docs, I believe if we update our App.MainWindow to inherit from WindowEx and then add a global using statement for the extension methods, that will enable core windowing functionality that most developers will want and need.

The extension methods make use of C#/Win32 to wrap P/Invokes to Win32 APIs, which will help bridge the gaps in the higher level platform.

### Additional context

Working with @dotmorten to make sure stability and quality are there to integrate into Template Studio.

### Applies to the following platforms:

- [X] WinUI
- [ ] WPF
- [ ] UWP

**Comment by mikebattista** [comment]
@dotmorten I integrated this pretty easily but am finding that the app process isn't exiting cleanly if I inherit from WindowEx or use any of the Window extension methods. Is this a known issue?

When debugging, the debugger never exits after closing the app and you can see the app process under Background Processes in Task Manager. If I start without debugging, every launch creates a new instance of the app, and you can see the old instances hanging around in Task Manager again under Background Processes after closing them. This also prevents rebuilding the app since VS can't overwrite the app binaries used by the running instances.

**Comment by dotMorten** [comment]
@mikebattista I've never seen this happen. Do you have a simple repro?

**Comment by mikebattista** [comment]
I'll publish a reference project with steps to add WinUIEx so you can see the before and after behavior.

**Comment by mikebattista** [comment]
Inheriting from WindowEx seems to trigger the repro. Just calling the CenterOnScreen extension method with the existing Window didn't seem to repro.

The reference project is https://github.com/mikebattista/TemplateStudio-WinUIEx. To run my repro, update [ActivateAsync](https://github.com/mikebattista/TemplateStudio-WinUIEx/blob/8873b39a1c33baa69de6807139b1a16ea2365678/TemplateStudio-WinUIEx/Services/ActivationService.cs#L24-L44) with the below.

    public async Task ActivateAsync(object activationArgs)
    {
        // Execute tasks before activation.
        await InitializeAsync();

        // Set the MainWindow Content.
        if (App.MainWindow is not WindowEx)
        {
            App.MainWindow = new WindowEx();
            _shell = App.GetService<ShellPage>();
            App.MainWindow.Content = _shell ?? new Frame();
        }

        // Handle activation via ActivationHandlers.
        await HandleActivationAsync(activationArgs);

        // Activate the MainWindow.
        App.MainWindow.Activate();

        App.MainWindow.CenterOnScreen();

        // Execute tasks after activation.
        await StartupAsync();
    }

I had to change the `App.MainWindow.Content == null` check since the WindowEx constructor [adds some stub Content](https://github.com/dotMorten/WinUIEx/blob/e1775289d7fb13cf622d12198113260f080b5f42/src/WinUIEx/WindowEx.cs#L63).

**Comment by mikebattista** [comment]
This might be related to the fact the App.xaml.cs is setting MainWindow to a new Window() as before and then I'm reassigning it to WindowEx above. That initial Window might be getting orphaned and causing problems that keep the process alive. If I remove that initial existing new Window() call, everything seems to work.

I'd just set App.xaml.cs to new WindowEx() but that sets the Content in the constructor which invalidated the existing conditional. I think if I leave MainWindow null in App.xaml.cs and check for null before initializing in ActivateAsync that should work.

**Comment by mikebattista** [comment]
The below would also work and avoids having to deal with Nullable elsewhere in the code.

```csharp
public static WindowEx MainWindow { get; set; } = new WindowEx() { Title = "AppDisplayName".GetLocalized(), Content = null };
```

**Comment by mikebattista** [comment]
While testing I noticed a few things:

* `IsMinimizable` and `IsMaximizable` have no effect for the custom title bar we use
* `SetIsMinimizable` and `SetIsMaximizable` also have no effect, and it's odd that there are properties and methods that are named similarly but do slightly different things (at least for the system title bar). 1 disables the buttons and the other removes them completely.
* Other functionality like IsAlwaysOnTop and IsResizable works great, but also they have both Get/Set methods as well as properties. Why the duplication?

**Comment by dotMorten** [comment]
Could it be you're looking at the extension methods from the WindowExtensions class, as well as first-class properties on WindowEx?

Btw I'm toying with the idea of taking all the "meat" of the WindowEx class and putting it into a helper that can be attached to any Window, instead of subclassing. Perhaps this would be cleaner for TS? Most of the work is already done, but the class is still internal and I'm toying with getting the API right.

**Comment by mikebattista** [comment]
> Could it be you're looking at the extension methods from the WindowExtensions class, as well as first-class properties on WindowEx?

I guess so. MainWindow is typed as WindowEx and I added a global using statement so the extension methods are available everywhere by default. You need to do both to get all the functionality so the combined view is what developers would see. Can we remove the duplication or at least adjust the naming to reflect the difference in behavior?

> Btw I'm toying with the idea of taking all the "meat" of the WindowEx class and putting it into a helper that can be attached to any Window, instead of subclassing. Perhaps this would be cleaner for TS? Most of the work is already done, but the class is still internal and I'm toying with getting the API right.

Interesting. So MainWindow can remain a Window and a global using statement will bring in all of the combined functionality of WindowEx and the extensions? What would "attaching to a Window" look like?

**Comment by dotMorten** [comment]
> Can we remove the duplication or at least adjust the naming to reflect the difference in behavior?

No that would be a major change in behavior. There shouldn't really be a difference in behavior, since internally the extension methods should be used (if there is, it's a bug)

> What would "attaching to a Window" look like?

**Comment by mikebattista** [comment]
Ok. If the recommendation is to use one or the other, then WindowEx seems to be the clear winner:

* Uses properties over methods which I prefer
* Enables Mica just by setting the Backdrop. Awesome!
* Remembers window size and position though I haven't seen this working yet.
  * Do I need to do something for window size and position to be remembered? I found PeristenceId but setting it didn't seem to have any effect.
* Automatic handling of DPI changes and resizes

That said, the extension methods have functionality that WindowEx doesn't expose:

* CenterOnScreen
* CreateMessageDialog (ShowMessageDialogAsync might be an equivalent)
* CreateOpenFilePicker
* CreateSaveFilePicker
* GetDpiForWindow
* GetWindowHandle
* Hide
* Maximize
* Minimize
* MoveAndResize
* Restore
* SetForegroundWindow (is WindowEx BringToFront equivalent?)
* SetIcon
* SetTitleBarBackgroundColors
* SetWindowSize (is setting WindowEx Width and Height equivalent?)
* Show

If we shouldn't use the extensions together with WindowEx, do you have a goal to bring this functionality to WindowEx for parity with the extensions? I added the global using so that all of this additional functionality would be provided but that of course resulted in the perceived duplication in Intellisense for the common functionality.

> There shouldn't really be a difference in behavior, since internally the extension methods should be used (if there is, it's a bug)

I retested the minimizable/maximizable APIs and they both hide the caption buttons now at least for the system title bar. I definitely saw the caption buttons disabled though at one point. Neither works though with the custom title bar we set with Window.ExtendsContentIntoTitleBar and Window.SetTitleBar. Do you think this could be a bug in WinUIEx or do we need to follow up with the owners of the custom title bar code that draws the caption buttons?

Regarding "attaching to a Window", for my needs and in terms of integrating with Template Studio, the existing WindowEx design is great. What would be the benefits of the helper approach?

**Comment by dotMorten** [comment]
This is your problem:
```
  if (App.MainWindow is not WindowEx)
    {
        App.MainWindow = new WindowEx();
```
You're replacing the first window with a new one, so the old one still stays around and WinUI thinks that means the app is still running.
Is there a reason you're not creating WindowEx [in app.xaml.cs](https://github.com/mikebattista/TemplateStudio-WinUIEx/blob/8873b39a1c33baa69de6807139b1a16ea2365678/TemplateStudio-WinUIEx/App.xaml.cs#L75)? 

Alternatively, make sure you close the old window (call `Close()`) after creating the the new window. But seems silly to create the first window when you're going to replace it anyway.

In any case, this issue isn't related to WinUIEx, but just how Windows are handled in WinUI3. You can still reproduce it if you create a "normal" window instance instead.

**Comment by mikebattista** [comment]
Right I'm not doing that anymore. I initially tried setting it in App.xaml.cs but ran into issues that I later found out were because you set the Content in the WindowEx constructor. I can initialize in App.xaml.cs if I set Content = null which is what I'm doing now.

I do have some questions [here](https://github.com/microsoft/TemplateStudio/issues/4529#issuecomment-1166589303) though.

Regarding Persistence I've since discovered that it's only supported for MSIX apps. Would be great if we could support unpackaged apps as well which is what I've been testing with.

I've also discovered that some of the missing functionality from the extension methods like Show/Hide/etc. is available on the WindowEx.AppWindow property. That's less discoverable and usable than if they were available on WindowEx itself. It looks like you already provide some passthrough APIs on WindowEx like IsShownInSwitchers and Presenter. Have you thought about applying this pattern to other AppWindow APIs as well?

**Comment by dotMorten** [comment]
> Would be great if we could support unpackaged apps as well which is what I've been testing with.

Unfortunately the APIs needed to store this are part of the packaged apps. Unpackaged apps doesn't have this ApplicationData available for storage. I'd be open to other ideas that "just work" preferably with out extra custom settings - feel free to open an enhancement require in the WinUIEx repo.

> I do have some questions here though.

I thought we covered these already? I do encourage you to log individual issues or open discussions in the WinUIEx repo - it's a little easier to manage that way.

**Comment by mikebattista** [comment]
I will log issues in your repo. Thanks.

Other than proposed enhancements, integration of WindowEx in Template Studio has been very easy and I think will add a lot of value.

**Comment by dotMorten** [comment]
Great to hear. FYI just published v1.7.0 which allows you to extend any window with WindowEx's features using the new 'WindowManager' class (it's internally what powers WindowEx).
